### ISSUE TITLE
[Discussion] Bind to specific IP

### ISSUE THREAD

**Issue by gelsbern** [citizenfx/txAdmin#21]
I have looked through everything and I would want to bind this to the specific IP of 127.0.0.1 and then proxy it behind nignx or apache.  Is there a way to do that?

**Comment by tabarra** [comment]
May I ask you why?  
Because if you want to do this reverse proxy for SSL support, then you don't need to do this anymore since we are rolling out SSL support today.

**Comment by gelsbern** [comment]
Not for SSL, but because I don't want it answering to all 17 of the IP addresses on the server.  Right now it binds to 0.0.0.0 which means no matter which IP is used if they connect to the 40120 they get to the txadmin server.  I want to confine it to a specific IP.  Proxying was the easiest method plus I can add in some added protections.

**Comment by tabarra** [comment]
For now, can you edit the source code for your needs?  
Edit `src/components/webServer.js` in the constructor.  
  
This can be easily fixed by adding a hidden setting for binding interface, but I just don't know if its worth adding this one more question to ask the inexperienced users in order to troubleshoot their problems.

**Comment by parkervcp** [comment]
Seeing as txAdmin is now built into cfxserver I don't see why having a convar to set this from the default would be an issue.

**Comment by tabarra** [comment]
I'm just not sure if this is worth the potential troubleshooting issues we will get.
a machine having 17 interfaces is as "edge case" as it gets.  
  
The only legitimate reason I see for this is if you have 2~3 servers, and want all of them to have the same txAdmin port.  
But for that, we enter into a new discussion of what is the best way to handle parallel servers to interfere with each other.  
  
I need to listen for more arguments other than "why not". Check out feature graveyard documentation, I tend to remove things that are not that important, just to prevent issues.    
Making things is easy-ish, maintaining and giving support for it is way harder, need to consider many things.  
Uncountable the number of times where a discussion in the help channel ended with me saying "txadmin binds to all interfaces", and it this stops being true is one more things I will need to ask while trying to troubleshoot the very common "plz help I cant access txadmin".  
  
**In conclusion:** I will leave this and #186 open while I wait for better arguments, and possibly even suggestions on how to solve the multi-server issue.

**Comment by Trsak** [comment]
I actually do not see any cons to that. Sice you do not have to use that convar, nothing can get wrong. When troubleshooting, users should provide all convars used.

On the other hand, you can just block txAdmin port on those IPs you dont want to be used (using `iptables` etc.). Using just one convar would be much faster and better solution tho, since you would not have to block the port for every new IP in the future.

**Comment by 2called-chaos** [comment]
I also prefer binding locally and frankly the vast majority of software has an option for that. Reasons for nginx is that my SSL setup is already there (LE certs coming from another server and reloading nginx automatically) and I like to address it portless aka on 443. And frankly I couldn't easily find information on how the integrated SSL functionality is set up but it also doesn't really make a big difference regarding my points.

I'm not too sure if this is quite related but what is currently preventing me from using txAdmin completely is the fact that I can't not have a 0.0.0.0 endpoint in my server config. I have multiple IPs on my server and whilst I can connect to one of those when binded like this I can't connect to it over the one I'd like it to for DNS reasons.

Not even talking about hosting two instances on the same ports on different IPs as I did with many other gameservers as to be able to just `connect <gamemode>.<domain>.tld` and `https://admin.<gm>.<domain>.tld` without any port hassles.

**Comment by tabarra** [comment]
### Decision:
After over a year with this discussion opened, and talking to other people, **I decided against it**.  
This is not a technical matter (the code in #186 is perfectly fine), the issue is the troubleshooting headache which is easily underestimated by everybody that is not in our Discord guild _every single day_ answering questions regarding how to connect to txAdmin.  
This decision also took into consideration that only a handful asked for this feature... ever!  
  
I sympathize with the points brought forward in this discussion, but for those, I suggest you guys build txAdmin from source or use Docker.  Also let's not forget that being a Gaming Server Provider is against FiveM's ToS.  
  
ps.: @2called-chaos your issue has nothing to do with what we are discussing. You can include `add_endpoint_tcp "127.0.0.1:30120"` in your `server.cfg`.

**Comment by 2called-chaos** [comment]
> ps.: @2called-chaos your issue has nothing to do with what we are discussing. You can include `add_endpoint_tcp "127.0.0.1:30120"` in your `server.cfg`.

Well yeah but then I can't connect to the server if it binds the ports locally only. As I said it wasn't quite related. My point is that I can't have (or rather want) a bind on 0.0.0.0 (for the fivem server) and when I bind to my specific IP txadmin doesn't let me use the config as it prompts be to add a tcp_endpoint line with 0.0.0.0 to it. To clarify I have a `add_endpoint_tcp 11.22.33.44:30120` and txadmin refuses to use that config.

I mean it's fine. You decided and I respect that. Even if I think that there should be a way (hidden for advanced users that don't typically nag support) to have more choice when it comes to binding. Be it fivem or txadmin. [But I'm also the king of options](https://user-images.githubusercontent.com/218011/103321883-16721300-4a3c-11eb-88c5-66882324fd62.png) so my  opinion might be skewed :)

**Comment by tabarra** [comment]
@2called-chaos  Well... i said "include" as in "add"... you can have more than one tcp endpoint you know!?  
```
add_endpoint_tcp "xxxxxxxxxxxx:30120"
add_endpoint_udp "xxxxxxxxxxxx:30120"
add_endpoint_tcp "127.0.0.1:30120"
```

You can totally just do this, so you hear both on the WAN and on the Loopback interface.

**Comment by 2called-chaos** [comment]
> @2called-chaos Well... i said "include" as in "add"... you can have more than one tcp endpoint you know!?

frankly I didn't. Thanks for the info, now I can give txadmin a proper try :) If it works that is because binding to 0.0.0.0 didn't let me connect to all my 3 IPs, just the first despite the port being open on all of them. But I will try when I have completed my server upgrade. Thanks again
